pkgs/dart_tooling_mcp_server/lib/src/mixins/dtd.dart

```diff
@@ -97,6 +99,95 @@ base mixin DartToolingDaemonSupport on ToolsSupport {
   // TODO: support passing a debug session id when there is more than one debug
   // session.
   Future<CallToolResult> takeScreenshot(CallToolRequest request) async {
+    return _callOnVmService(
+      callback: (vmService) async {
+        final vm = await vmService.getVM();
+        final result = await vmService.callServiceExtension(
+          '_flutter.screenshot',
+          isolateId: vm.isolates!.first.id,
+        );
+        if (result.json?['type'] == 'Screenshot' &&
+            result.json?['screenshot'] is String) {
+          return CallToolResult(
+            content: [
+              ImageContent(
+                data: result.json!['screenshot'] as String,
+                mimeType: 'image/png',
+              ),
+            ],
+          );
+        } else {
+          return CallToolResult(
+            isError: true,
+            content: [
+              TextContent(
+                text: 'Unknown error or bad response taking screenshot:\n'
+                    '${result.json}',
+              ),
+            ],
+          );
+        }
+      },
+    );
+  }
+
+  /// Performs a hot reload on the currently running app.
+  ///
+  /// If more than one debug session is active, then it just uses the first one.
+  ///
+  // TODO: support passing a debug session id when there is more than one debug
+  // session.
+  Future<CallToolResult> hotReload(CallToolRequest request) async {
+    return _callOnVmService(
+      callback: (vmService) async {
+        final vm = await vmService.getVM();
+
+        String? hotReloadMethodName;
+        vmService.onEvent(EventStreams.kService).listen((Event e) {
+          if (e.kind == EventKind.kServiceRegistered) {
+            final serviceName = e.service!;
+            if (serviceName == 'reloadSources') {
+              // This may look something like 's0.reloadSources'.
+              hotReloadMethodName = e.method!;
+            }
+          }
+        });
+        await vmService.streamListen(EventStreams.kService);
+
+        // Short delay to allow the service registration event to be sent.
+        await Future<void>.delayed(const Duration(seconds: 1));
```

Comment:
nit: make the hotReloadMethodName a completer and complete it in the onEvent listener, then await that here with a timeout